node/cmd/spy/spy.go

```diff
@@ -136,8 +141,198 @@ func (s *spyServer) Publish(vaaBytes []byte) error {
 	return nil
 }
 
+// TransactionIdMatches decodes both transactionIDs and checks if they are the same.
+func TransactionIdMatches(batch *vaa.BatchVAA, t *spyv1.BatchFilter) (matches bool, err error) {
+
+	// first check if the transaction IDs match
+	filterHash, err := vaa.StringToHash(t.TransactionId)
+	if err != nil {
+		return matches, status.Error(codes.InvalidArgument, fmt.Sprintf("failed to decode filter's txId: %v", err))
+	}
+
+	matches = filterHash == batch.TransactionID
+	return matches, nil
+}
+
+// BatchMatchFilter asserts that the obervation matches the values of the filter.
+func BatchMatchesFilter(batch *vaa.BatchVAA, f *spyv1.BatchFilter) (matches bool, err error) {
+
+	txMatch, err := TransactionIdMatches(batch, f)
+	if err != nil {
+		return matches, err
+	}
+
+	if txMatch {
+		// the BatchVAA's transaction ID matches the transaction ID of this filter.
+		// now check if the other properties of the filter match.
+
+		if obs := batch.Observations[0]; obs != nil {
+			obsVAA := obs.Observation
+
+			if obsVAA.EmitterChain == vaa.ChainID(f.ChainId) {
+				// the emitter chain of the observation matches the filter
+
+				if f.Nonce >= 1 {
+					// filter has a nonce, so make sure it matches
+					if obsVAA.Nonce == f.Nonce {
+						// filter's nonce matches the nonce of the obervations. send it.
+						matches = true
+						return matches, err
+					}
+
+				} else {
+					// filter does not have a nonce, everything else matched, send it.
+					matches = true
+					return matches, err
+				}
+			}
+		}
+	}
+	return matches, err
+}
+
+func (s *spyServer) PublishSignedVAAByType(vaaBytes []byte) error {
+	s.subsAllVaaMu.Lock()
+	defer s.subsAllVaaMu.Unlock()
+
+	// this will try to unmarshal the byte array to a VAA, and then to a BatchVAA.
+	// the unmarshaling could populate one of the variable below. it's also possible
+	// a new type of vaaBytes comes through, so the variables below could remain empty.
+
+	var v *vaa.VAA
+	var b *vaa.BatchVAA
+
+	v, err := vaa.Unmarshal(vaaBytes)
+	// do nothing with the error, until we can try to unmarshal the bytes as a batch.
+	if err != nil {
+		// check if it is a batch
+
+		// it is not a VAA, try unmarshaling to a BatchVAA
+		b, err = vaa.UnmarshalBatch(vaaBytes)
+		if err != nil {
+			// it is not either type of VAA we know, nothing to do.
+			return err
+		}
+	}
+
+	// create the response(s) that will get sent out if this VAA satisfies a subscription.
+
+	// create the top-level response struct that is agnostic to the VAA type
+	var topRes *spyv1.SubscribeSignedVAAByTypeResponse
+
+	// if v has values, the VAA unmarshal was successful.
+	if len(v.Payload) > 0 {
```

Comment:
Can't v be nil?

Reply:
I refactored this function into separate functions for the different VAA types, it's much cleaner, and it gets rid this hacky "is struct populated" check.